clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -228,7 +229,8 @@ private void poll(long timeout, long now, boolean executeDelayedTasks) {
         trySend(now);
 
         // fail all requests that couldn't be sent
-        failUnsentRequests();
+        if (failUnsent)
+            failUnsentRequests();
```

Comment:
Rather than trying to control when and when not to fail unsent requests, I wonder if we should use "request.timeout.ms" to selectively fail only those requests whose timeouts have expired? Then we could remove the `failUnsent` flag everywhere and just call something like `timeoutExpiredRequests()`. In that case, we could remove SendFailedException entirely and use TimeoutException. What do you think?

The only problem is that I'm not sure whether this could cause strange behavior on coordinator fail-over. For example, suppose that we try to send a commit just before the coordinator crashes. In that case, the commit could get stuck in the request queue. We may find the new coordinator and the old coordinator may even come back before the commit's timeout has expired. In that case, if the commit eventually is sent to the old coordinator, then we'd suddenly see a NOT_COORDINATOR_FOR_GROUP error, which would cause us to find the coordinator again unnecessarily. Maybe all we need to address edge cases like this is a way to fail any pending requests to the coordinator when we find that it has failed.